x/qgb/abci.go

```diff
@@ -113,3 +122,118 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
+		return
+	}
+	// now we want to prune attestations as we have the following conditions:
+	// - we have attestations up to the last unbonding height
+	// - the total number of attestations, including the ones before the unbonding height, are greater
+	// than the AttestationPruningThreshold
+	err := pruneAttestations(ctx, k, lastAvailableNonce, latestAttestationNonce)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// getLastAvailableNonceHeight returns the block height corresponding to the last available
+// nonce in store.
+func getLastAvailableNonceHeight(ctx sdk.Context, k keeper.Keeper) uint64 {
+	lastAttestationInStore, found, err := k.GetAttestationByNonce(ctx, k.GetLastAvailableAttestationNonce(ctx))
+	if err != nil {
+		panic(err)
+	}
+	if !found {
+		panic(fmt.Sprintf("couldn't find attestation %s for pruning", lastAttestationInStore))
+	}
+	switch lastAttestationInStore.Type() {
+	case types.DataCommitmentRequestType:
+		dc := lastAttestationInStore.(*types.DataCommitment)
+		return dc.EndBlock
+	case types.ValsetRequestType:
+		vs := lastAttestationInStore.(*types.Valset)
+		return vs.Height
+	default:
+		panic("unknown attestation type")
+	}
+}
+
+// pruneAttestations prunes attestations to keep attestations up to the last unbonding height
+// and the total number of attestations lower than the AttestationPruningThreshold.
+func pruneAttestations(
+	ctx sdk.Context,
+	k keeper.Keeper,
+	lastAvailableAttestationNonce uint64,
+	latestAttestationNonce uint64,
+) error {
+	ctx.Logger().Debug("pruning attestations from qgb store")
+	if !k.CheckLastUnbondingNonce(ctx) {
+		// We should never hit this case in the happy path because at this level, we're sure that
+		// all store values are initialized.
+		// However, we might hit it in an upgrade if we don't handle the store values carefully.
+		// So, it's good to have this check in place.
+		return fmt.Errorf("last unbonding nonce not initialized in store")
+	}
+	lastUnbondingNonce := k.GetLastUnbondingNonce(ctx)
+	count := 0
+	newLastAvailableNonce := lastAvailableAttestationNonce
+	for i := lastAvailableAttestationNonce; i < lastUnbondingNonce; i++ {
+		if newLastAvailableNonce == lastUnbondingNonce ||
+			// the +1 is because we will have the following range in store:
+			// [lastAvailableAttestationNonce, latestAttestationNonce] with its
+			// length == AttestationPruningThreshold
+			latestAttestationNonce-i+1 == AttestationPruningThreshold {
+			// we either reached the last unbonding height, and we need to keep all those attestations
+			// or, we reached the minimum number of attestations we want to keep in store
+			newLastAvailableNonce = i
+			break
+		}
+		k.DeleteAttestation(ctx, i)
+		count++
+	}
+	// persist the new last available attestation nonce
+	k.SetLastAvailableAttestationNonce(ctx, newLastAvailableNonce)
+	ctx.Logger().Debug(
+		"finished pruning attestations from qgb store",
+		"count",
+		count,
```

Comment:
```suggestion
        // loop through all the attestations between the last available nonce and the last nonce within the unbonding period
	for newLastAvailableNonce := lastAvailableAttestationNonce; newLastAvailableNonce < lastUnbondingNonce;  {
	        // check how many attestations we have
		if latestAttestationNonce-newLastAvailableNonce+1 == AttestationPruningThreshold {
			// we have reached the minimum number of attestations we want to keep in store
			// we don't want to prune any more, so we break
			break
		}
		k.DeleteAttestation(ctx, newLastAvailableNonce)
		newLastAvailableNonce++
	}
	// persist the new last available attestation nonce
	k.SetLastAvailableAttestationNonce(ctx, newLastAvailableNonce)
	ctx.Logger().Debug(
		"finished pruning attestations from qgb store",
		"count",
		newLastAvailableNonce-lastAvailableAttestationNonce,
```
nit: I think this is a little more readable

Reply:
[dfa4bbe](https://github.com/celestiaorg/celestia-app/pull/1799/commits/dfa4bbeabad638d6572728db36766a04959c7a41)